LongWarOfTheChosen/Src/NewPromotionScreenByDefault_Integrated/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -309,127 +309,146 @@ function bool HasBrigadierRank()
 		}
 		else
 		{
-			Column.AbilityNames.AddItem(''); // Make sure we add empty spots to the name array for getting ability info
-			Column.AbilityIcons[iAbility - Position].Hide();
-			Column.InfoButtons[iAbility - Position].Hide();
-			Column.MC.ChildSetBool("EquippedAbility" $ (iAbility - Position), "_visible", false);
+			// Make sure we add empty spots to the name array for getting ability info.
+			Column.AbilityNames.AddItem('');
+			Column.AbilityIcons[AbilityIndex - Position].Hide();
+			Column.InfoButtons[AbilityIndex - Position].Hide();
+			Column.MC.ChildSetBool("EquippedAbility" $ (AbilityIndex - Position), "_visible", false);
 		}
 	}
 
-	// bsg-nlong (1.25.17): Select the first available/visible ability in the column
-	// NPSBDP: It is possible for ranks to have no abilities if we offset them in a way that hides all ability icons
-	// So only do this if we have visible ability icons
-	while(`ISCONTROLLERACTIVE && !AllAbilityIconsHidden(Column) && !Column.AbilityIcons[Column.m_iPanelIndex].bIsVisible)
+	// XCom : If using a controller, select the first available/visible ability in the column.
+	// NPSBDP : It is possible for ranks to have no abilities if we offset them in a way that hides all ability icons;
+	// therefore, only do this if we have visible ability icons.
+	while (`ISCONTROLLERACTIVE && !AllAbilityIconsHidden(Column) && !Column.AbilityIcons[Column.m_iPanelIndex].bIsVisible)
 	{
 		Column.m_iPanelIndex +=1;
-		if( Column.m_iPanelIndex >= Column.AbilityIcons.Length )
+		if (Column.m_iPanelIndex >= Column.AbilityIcons.Length)
 		{
 			Column.m_iPanelIndex = 0;
 		}
 	}
-	// bsg-nlong (1.25.17): end
 
-	return bHasColumnAbility;
+	// KDM : The return value, HasColumnAbility, helps determine whether this column should have the 
+	// 'New Rank' graphic shown over top of its rank name and symbol. Unfortunately, the code above only
+	// checks whether a 'visible' ability has been chosen in this given column; it ignores abilities
+	// which are invisible due to scrolling. 
+	//
+	// Here is a simple example. Let's say we scroll down and are currently viewing the bottom 2 soldier
+	// ability rows, the XCom ability row, and the pistol ability row. Furthermore, let's assume that we 
+	// have already purchased the ability, within this column, located in the top soldier ability row. 
+	// Although HasColumnAbility should be true, due to having an ability in this column, it will, in fact, 
+	// be false since the invisible ability row is never checked.
+	//
+	// What we really want to do is, for this particular column, determine if one of its 'soldier abilities' 
+	// have been chosen/bought while, at the same time, ignoring 'XCom abilities' and 'pistol abilities'. 
+	// Since 'soldier abilities' are located in the top 3 rows, we simply need to see if any of them have been 
+	// chosen/bought regardless of visibility.
+	HasColumnAbility = false;
+	if (!ColumnAboveUnitRank)
+	{
+		for (i = 0; i < 3; i++)
```

Comment:
What's the 3 for here? If this is the number of class abilities available at each rank, then the value needs to be pulled from `GetAbilitiesPerRank(UnitState)`.

Reply:
I believe the 'New Rank' graphic should only appear if someone has not chosen a 'soldier ability' at their given rank. It should ignore whether or not an 'XCom' or 'Pistol' ability has been chosen at this rank.

- The 1st, 2nd, and 3rd abilities for a given rank correspond to the 1st, 2nd, and 3rd rows in rank column; these are the 'soldier abilities'. 
- The 4th ability for a given rank corresponds to the 4th row in a rank column, which is the 'XCom' ability. 
- The 5th ability, if it exists, for a given rank corresponds to the 5th row in a rank column, which is the 'Pistol' ability.

This checks to see if any 'soldier abilities' have been selected at their given rank.

Another way to do it would be to cycle through all class abilities, at a given rank, and only check 'soldier abilities'. Is there a way to determine if an ability is a 'soldier ability' vs 'XCom ability' vs 'Pistol ability'. This way ordering would be unimportant.

Reply:
See `GetAbilitiesPerRank()`. It's important not to use 3 because some classes may have only 2 or maybe more than 4. Depends on what soldier classes are added by mods.

Reply:
Right. That should give the total number of abilities for each rank; however, is there any way to find out "the type" of ability being looked at? 

Let's say I'm looking at a Sharpshooter at Lance Corporal rank. I can see 3 soldier abilities : _Death from Above_, _Rapid Targeting_, and _Snap Shot_. Furthermore, I see 1 XCom ability, _Flashbanger_, and 1 pistol ability, _Return Fire_.

For this modification to work properly, I would need to be able to determine if an ability is a base soldier ability or "some other" ability.

I believe in the base game the 'New Rank' graphic would show whenever an ability had not been chosen for the soldier's rank. However, things are more complicated in LW. Looking at the example above, let's say the pistol ability _Return Fire_ was bought; I still think 'New Rank' should show since a base soldier ability has not been bought. XCom abilities and pistol abilities seem like 'extras' but not 'core' upgrades.

Worst case, if this is not possible, I can just loop over "all" abilities like mentioned previously, and see if any have been bought.

Reply:
Yeah, sorry, when I glanced at it earlier I thought it was giving just the class abilities. I think it might be best to rely on the promotion UI config that says how many class abilities there are. `ClassAbilitiesPerRank`. If soldier class mods want to support the new promotion UI/LWOTC, they will need to add entries to that config array. I think I'm using it elsewhere already.

Reply:
Can we assume that class abilities always come before XCom/Pistol/Non-Core abilities? I'm guessing that class abilities will always take up the first 'n' rows while 'other' abilities will begin at row 'n+1'.

Reply:
That's the assumption I've worked on before. It's not foolproof, because I'm pretty sure it depends on the order in which they're declared in XComClassData.ini. Also, this probably wouldn't work at all for RPGO, but that's fine. I think just having a rule that LWOTC soldier classes have their class abilities first should work out OK.

Reply:
Within XComClassData.ini, I think a lot of AbilitySlots have an extra 'none' ability at the end.

For example : 
```
+SoldierRanks=(AbilitySlots=((AbilityType=(AbilityName="RevivalProtocol", ApplyToWeaponSlot=eInvSlot_SecondaryWeapon)), \\
(AbilityType=(AbilityName="Sentinel_LW", ApplyToWeaponSlot=eInvSlot_PrimaryWeapon)), \\
(AbilityType=(AbilityName="CombatProtocol", ApplyToWeaponSlot=eInvSlot_SecondaryWeapon)), \\
(RandomDeckName="Tier1_XComAbilities"), \\
(AbilityType=(AbilityName="ReturnFire", ApplyToWeaponSlot=eInvSlot_Pistol)), \\
***_THE_COMMA_JUST_BEFORE_THIS_STATEMENT_CREATES_AN_EXTRA_ELEMENT***), \\
aStatProgression=((StatType=eStat_Offense,StatAmount=2), (StatType=eStat_Will,StatAmount=5), (StatType=eStat_Hacking,StatAmount=10), (StatType=eStat_HP,StatAmount=1), (StatType=eStat_Strength,StatAmount=0), (StatType=eStat_CombatSims,StatAmount=0))\\
)
```

I noticed that when you log `AbilityTree.Length` after setting it via :
`AbilityTree = Unit.GetRankAbilities(Column.Rank);`
you often get 6 even though there are 5 abilities for a given rank.

I'm not sure if this matters or not. I'm assuming these 'none' abilities are just skipped in code generally speaking.

I will post it in issues and it can be closed if it is a non-issue.